### PR TITLE
chore: remove `unsafe`-s; comment `unwrap`-s and `numtoa_str`-s

### PR DESCRIPTION
**Pull request by dj8yfo**
# Checklist

- [ ] App update process has been followed 
- [x] Target branch is `develop` 
- [ ] Application version has been bumped

### CODE REVIEW

File: src/parsing/types/nep413/payload.rs [hunk 1]
```diff
@@ -18,6 +18,7 @@ impl Payload {
             callback_url: Some(CappedString::new()),
         }
     }
+    /// must be only called once after `Self::new` to avoid unexpected panics
```
Comment (frol) [inline src/parsing/types/nep413/payload.rs:21]: Is there a way to avoid unexpected panics or forbid subsequent calls through ownership interface?

**Review by frol — CHANGES_REQUESTED** [review]

File: src/app_ui/sign/common/action/add_key_common.rs [hunk 2]
```diff
@@ -29,34 +29,27 @@ pub fn format<'b, 'a: 'b, const N: usize>(
...
-                .numtoa_str(10, &mut field_context.num_buf),
-        }))
-        .unwrap();
+    writer.push_fields(ElipsisFields::one(Field {
+        name: "Access Key Nonce",
+        value: add_key
+            .access_key
+            .nonce
+            // numtoa_buf has to be at least 20 bytes for u64 (8 bytes) : ok
+            .numtoa_str(10, &mut field_context.num_buf),
```
[the remark this answers is not in the document]
Comment (frol) [reply src/app_ui/sign/common/action/add_key_common.rs:48]: It turns out that `10` is the base, not the length. Disregard my question.

[the remark this answers is not in the document]
Comment (dj8yfo) [reply src/app_ui/sign/common/action/add_key_common.rs:48]: no

File: src/utils/types/capped_string/hex_display.rs [hunk 3]
```diff
@@ -22,11 +35,15 @@ impl<const N: usize> HexDisplay<N> {
...
-    #[allow(unused)]
     pub fn truncated(&self) -> bool {
         self.truncated
     }
 
+    pub fn leftover(&self) -> usize {
+        self.leftover
+    }
+
+    /// this method must only be called once
```
[the remark this answers is not in the document]
Comment (frol) [reply src/utils/types/capped_string/hex_display.rs:46]: Just for the future context: we tried to fix it, but we operate with beefy inline structs and run out of stack memory, so there is no solution here yet.

Here is the code snippet that we tested:

```rust
impl BorshDeserialize for Payload {

    #[inline(always)]
    fn deserialize_reader<R: Read>(reader: &mut R) -> Result<Self> {
        let mut result = Ok(Self {
            message: CappedString::new(),
            nonce: [0u8; 32],
            recipient: CappedString::new(),
            callback_url: Some(CappedString::new()),
        });

        if let Ok(result) = result.as_mut() {

            result.message.deserialize_reader_in_place(reader)?;

            let nonce: [u8; 32] = BorshDeserialize::deserialize_reader(reader)?;
            result.nonce = nonce;

            result.recipient.deserialize_reader_in_place(reader)?;

            let option_flag: u8 = BorshDeserialize::deserialize_reader(reader)?;

            if option_flag == 0 {
                result.callback_url = None;
            } else if option_flag == 1 {
                // .unwrap() is ok if `result.callback_url` is Some
                result.callback_url
                    .as_mut()
                    .unwrap()
                    .deserialize_reader_in_place(reader)?;
            } else {
                return Err(Error::from(ErrorKind::InvalidData));
            }
          
        }

        result
    }
}
```

My assumption was that if we create the `Ok(Payload {})` and feel it in-place, we should be fine, but even that did not help.

File: src/parsing/types/nep413/payload.rs (hunk 1 above)
Comment (dj8yfo) [reply src/parsing/types/nep413/payload.rs:21]: https://github.com/dj8yfo/app-near-rs/pull/46#discussion_r1519872552

File: src/utils/types/capped_string/hex_display.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dj8yfo) [reply src/utils/types/capped_string/hex_display.rs:46]: we've tried this as an alternative
```rust
    // #[inline(always)]
    pub fn reformat(mut self) -> FormattedHexDisplay<N> {
        let prev_used = self.used;
        let mut new_used = self.used * 2;
        if new_used > self.buffer.len() {
            new_used = self.buffer.len()
        }

        self.used = new_used / 2;

        self.leftover += prev_used - self.used;
        if self.leftover > 0 {
            self.truncated = true;
        }
        let mut tmp_buffer = [0u8; 1];

        for ind in (0..self.used).rev() {
            let char_range = ind * 2..=ind * 2 + 1;
            tmp_buffer.copy_from_slice(&self.buffer[ind..ind + 1]);
            hex::encode_to_slice(tmp_buffer, &mut self.buffer[char_range]).unwrap();
        }
        FormattedHexDisplay {
            buffer: self.buffer,
            used: self.used,
            truncated: self.truncated,
            leftover: self.leftover,
        }
    }
```
```rust
    {
        // '{' char
        Some(123) => {
            let mut args_str: CappedString<200> = CappedString::new();
            match args_str.deserialize_with_bytes_count(stream, args_bytes_count) {
                Err(err) if err.kind() == ErrorKind::InvalidData => {
                    let args_bin: HexDisplay<200> = args_str.into();
                    ArgsRepr::BinHex(args_bin.reformat())
                }
                Ok(_) => ArgsRepr::String(args_str),
                Err(_err) => {
                    return Err(AppSW::TxParsingFail);
                }
            }
        }
        Some(_first_byte) => {
            let mut args_bin: HexDisplay<200> = HexDisplay::new();
            args_bin
                .deserialize_with_bytes_count(stream, args_bytes_count)
                .map_err(|_err| AppSW::TxParsingFail)?;

            ArgsRepr::BinHex(args_bin.reformat())
        }
        None => {
            return Err(AppSW::TxParsingFail);
        }
    };

```

File: src/utils/types/capped_string/hex_display.rs
[the remark this answers is not in the document]
Comment (dj8yfo) [reply src/utils/types/capped_string/hex_display.rs:1]: addressed by https://github.com/dj8yfo/app-near-rs/pull/46/commits/81726e6420632d0b703ce739e13270c76f611aee

File: src/utils/crypto/path.rs [hunk 4]
```diff
@@ -17,6 +17,8 @@ impl PathBip32 {
         }
 
         for (i, chunk) in data.chunks(mem::size_of::<u32>()).enumerate() {
+            // .unwrap() is ok, as `chunk.len() == 4` holds true
+            // https://doc.rust-lang.org/std/primitive.array.html#impl-TryFrom%3C%26%5BT%5D%3E-for-%5BT;+N%5D
             result.0[i] = u32::from_be_bytes(chunk.try_into().unwrap());
```
[the remark this answers is not in the document]
Comment (dj8yfo) [reply src/utils/crypto/path.rs]: addressed by https://github.com/dj8yfo/app-near-rs/pull/46/commits/c48493fc94f7bb14b4cfb3af12a4d80067e2443a

File: src/utils/crypto/public_key.rs [hunk 5]
```diff
@@ -55,8 +55,10 @@ impl PublicKeyBe {
     }
 
     pub fn display_str_hex<'b>(&self, buffer: &'b mut [u8; 64]) -> &'b str {
```
[the remark this answers is not in the document]
Comment (dj8yfo) [reply src/utils/crypto/public_key.rs]: addressed by https://github.com/dj8yfo/app-near-rs/pull/46/commits/5d8c7886224f8db575a1323ac08eb8c6cc6d2e5e

File: src/utils/types/base58_buf.rs [hunk 6]
```diff
@@ -15,14 +15,16 @@ impl<const N: usize> Base58Buf<N> {
         self.len = 0;
 
         // expecting `bs58` to always produce correct strings
-        // https://docs.rs/bs58/0.5.0/src/bs58/encode.rs.html#201
         let len = bs58::encode(target).onto(&mut self.buf[..])?;
```
[the remark this answers is not in the document]
Comment (dj8yfo) [reply src/utils/types/base58_buf.rs]: addressed by https://github.com/dj8yfo/app-near-rs/pull/46/commits/cb97684ebc168051c53e15440438e2eb0e661475

**Review by frol — APPROVED** [review]

**Pull request merged**